### PR TITLE
Added base64 helper

### PR DESCRIPTION
**Pull request by freekode**
# What does this implement/fix?

Added encode and decode funcitons for base64
Copied from [here](https://stackoverflow.com/questions/180947/base64-decode-snippet-in-c)

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes <link to issue>

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2935

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266
- [ ] RP2040

## Example entry for `config.yaml`:

```yaml
esp32_camera:
  ...
  on_image:
      then:
        - lambda: |-
            const char* image_data_char = reinterpret_cast<char*>(image.data);
            std::string encodedData = base64_encode(&image_data_char[0], image.length);
            bool result = id(mqtt_client).publish("cam/image", encodedData);
```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

**Comment by freekode** [comment]
I'm not sure which tests to add.
I tested it locally only.

**Comment by max246** [comment]
Maybe would be good to add a note in the esphome doc? If it gets approved is handy to know that i can be used for espcamera or just generic encoding.

**Comment by freekode** [comment]
Sure thing. I added explanation to esp32 camera docs. If you think there is another better place, jut let me know.

**Review by kroimon — APPROVED** [review]
Looks good to me!

File: esphome/core/helpers.h [hunk 1]
```diff
@@ -429,6 +429,12 @@ std::string value_accuracy_to_string(float value, int8_t accuracy_decimals);
 /// Derive accuracy in decimals from an increment step.
 int8_t step_to_accuracy_decimals(float step);
 
+static const std::string BASE64_CHARS = "ABCDEFGHIJKLMNOPQRSTUVWXYZ"
+                                        "abcdefghijklmnopqrstuvwxyz"
+                                        "0123456789+/";
+std::string base64_encode(const char *buf, unsigned int buf_len);
```
Comment (jesserockz) [inline esphome/core/helpers.h]:
```suggestion
std::string base64_encode(const uint8_t *buf, size_t buf_len);
```

Comment (jesserockz) [reply esphome/core/helpers.h]: Might be nice to also create overloads for these to they are consistent, ie decoding can take a buffer to write to and return the length as other buffer apis do.

And an overload to pass a std::vector to `encode` as well

**Review by jesserockz — CHANGES_REQUESTED** [review]

**Review by jesserockz — APPROVED** [review]

**Pull request merged**